[SPARK-24103][ML][MLLIB] ML Evaluators should use weight column - added weight column for binary classification evaluator

imatiach-msft opened the pull request:
## What changes were proposed in this pull request?

The evaluators BinaryClassificationEvaluator, RegressionEvaluator, and MulticlassClassificationEvaluator and the corresponding metrics classes BinaryClassificationMetrics, RegressionMetrics and MulticlassMetrics should use sample weight data.

I've closed the PR: https://github.com/apache/spark/pull/16557
as recommended in favor of creating three pull requests, one for each of the evaluators (binary/regression/multiclass) to make it easier to review/update.

## How was this patch tested?
I added tests to the metrics and evaluators classes.

imatiach-msft committed:
Added weight column for binary classification evaluator

SparkQA commented:
**[Test build #73526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73526/testReport)** for PR 17084 at commit [`98652cf`](https://github.com/apache/spark/commit/98652cfb6c92eed90deff61bc83ef66b9096df20).

SparkQA commented:
**[Test build #73526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73526/testReport)** for PR 17084 at commit [`98652cf`](https://github.com/apache/spark/commit/98652cfb6c92eed90deff61bc83ef66b9096df20).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BinaryClassificationMetrics @Since(\"2.2.0\") (`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73526/
Test PASSed.

imatiach-msft commented:
@sethah @Lewuathe @thunterdb @WeichenXu123 @jkbradley @actuaryzhang @srowen would you be able to take a look?  I've split the larger pull request into three parts as suggested.

imatiach-msft commented:
ping @sethah @Lewuathe @thunterdb @WeichenXu123 @jkbradley @actuaryzhang @srowen could you please take a look?  thank you!

actuaryzhang commented:
Thanks for the PR. I think this is helpful. Will take a look next week. Quite swamped recently.

HyukjinKwon commented:
gentle ping @actuaryzhang

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/BinaryClassificationEvaluator.scala:
```diff
@@ -77,12 +87,16 @@ class BinaryClassificationEvaluator @Since("1.4.0") (@Since("1.4.0") override va
-        case Row(rawPrediction: Double, label: Double) => (rawPrediction, label)
+    val scoreAndLabelsWithWeights =
+      dataset.select(col($(rawPredictionCol)), col($(labelCol)).cast(DoubleType),
+        if (!isDefined(weightCol) || $(weightCol).isEmpty) lit(1.0) else col($(weightCol)))
```
Check weightCol is double?

imatiach-msft replied:
added check for numeric type and did cast to Double, similar to labelCol

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/BinaryClassificationEvaluator.scala:
```diff
@@ -36,12 +36,18 @@ import org.apache.spark.sql.types.DoubleType
   /**
+   * Default number of bins to use for binary classification evaluation.
+   */
+  val defaultNumberOfBins = 1000
```
Why 1000?

srowen replied:
I think this should be a default in BinaryClassificationMetrics then?

srowen replied:
I'm not sure this was resolved; can the metrics class just use this as a default?

imatiach-msft replied:
done, made it default in BinaryClassificationMetrics

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -41,13 +41,27 @@ import org.apache.spark.sql.DataFrame
+   * Retrieves the score and labels (for binary compatibility).
+   * @return The score and labels.
+   */
+  @Since("1.0.0")
```
inconsistent annotation. was 1.3.0

imatiach-msft replied:
good catch, updated version for both:
1.) def scoreAndLabels: RDD[(Double, Double)] = ....
2.) def this(@Since("1.3.0") scoreAndLabels: RDD[(Double, Double)], @Since("1.3.0") numBins: Int) = ...
to 1.3.0

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryConfusionMatrix.scala:
```diff
@@ -22,22 +22,22 @@ package org.apache.spark.mllib.evaluation.binary
 private[evaluation] trait BinaryConfusionMatrix {
   /** number of true positives */
-  def numTruePositives: Long
+  def numTruePositives: Double
```
I feel it may be better to create new attributes like `weightedTruePositives`

imatiach-msft replied:
good idea, updated the names of the variables

actuaryzhang commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -146,11 +160,13 @@ class BinaryClassificationMetrics @Since("1.3.0") (
+    // negatives within each bin, and then sort by score values in descending order.
+    val counts = scoreAndLabelsWithWeights.combineByKey(
+      createCombiner = (labelAndWeight: (Double, Double)) =>
+        new BinaryLabelCounter(0L, 0L) += (labelAndWeight._1, labelAndWeight._2),
```
`new BinaryLabelCounter(0.0, 0.0)`? Defined to take double parameters below.

imatiach-msft replied:
updated, thanks!

actuaryzhang commented:
@imatiach-msft Thanks for the PR. Added a couple of comments. Sorry for the delayed review.

HyukjinKwon commented:
gentle ping @imatiach-msft .

imatiach-msft commented:
yes, will update the PR, thanks for the ping

SparkQA commented:
**[Test build #78597 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78597/testReport)** for PR 17084 at commit [`cf59c62`](https://github.com/apache/spark/commit/cf59c62f272ade192dfbf28ab53881251ea0d95e).

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/BinaryClassificationEvaluator.scala:
```diff
@@ -36,12 +36,18 @@ import org.apache.spark.sql.types.DoubleType
   /**
+   * Default number of bins to use for binary classification evaluation.
+   */
+  val defaultNumberOfBins = 1000
```
It seemed like a good default value to use - for graphing ROC curve, it's not too large for most plots, but it's not so small that the graph would be jagged.  The user can always specify a value to override the default.  However, it's usually not a good idea to sort over the entire label/score values, since the dataset will probably be very large, the operation will be very slow, and when visualizing the data there won't be any difference, so by default we should try to discourage the user from not down-sampling the number of bins.

imatiach-msft committed:
Updated based on comments - fixed since tag, renamed vars, added check for weight col

imatiach-msft commented:
Jenkins, test this please

SparkQA commented:
**[Test build #78598 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78598/testReport)** for PR 17084 at commit [`60fc2a7`](https://github.com/apache/spark/commit/60fc2a78d4c3e985e91fd14522642d861df58d99).

SparkQA commented:
**[Test build #78597 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78597/testReport)** for PR 17084 at commit [`cf59c62`](https://github.com/apache/spark/commit/cf59c62f272ade192dfbf28ab53881251ea0d95e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BinaryClassificationMetrics @Since(\"2.2.0\") (`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78597/
Test PASSed.

SparkQA commented:
**[Test build #78598 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78598/testReport)** for PR 17084 at commit [`60fc2a7`](https://github.com/apache/spark/commit/60fc2a78d4c3e985e91fd14522642d861df58d99).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78598/
Test FAILed.

imatiach-msft commented:
the pip packaging failing seems to be unrelated to the code... let me try this again

imatiach-msft commented:
Jenkins, test this please

SparkQA commented:
**[Test build #78606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78606/testReport)** for PR 17084 at commit [`60fc2a7`](https://github.com/apache/spark/commit/60fc2a78d4c3e985e91fd14522642d861df58d99).

SparkQA commented:
**[Test build #78606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78606/testReport)** for PR 17084 at commit [`60fc2a7`](https://github.com/apache/spark/commit/60fc2a78d4c3e985e91fd14522642d861df58d99).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78606/
Test PASSed.

imatiach-msft commented:
gentle ping @actuaryzhang, thanks!

imatiach-msft commented:
Jenkins, test this please

SparkQA commented:
**[Test build #89405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89405/testReport)** for PR 17084 at commit [`e89a030`](https://github.com/apache/spark/commit/e89a030aa63945ddda22e2f879e9e0cafebe0801).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2350/
Test PASSed.

SparkQA commented:
**[Test build #89405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89405/testReport)** for PR 17084 at commit [`e89a030`](https://github.com/apache/spark/commit/e89a030aa63945ddda22e2f879e9e0cafebe0801).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89405/
Test PASSed.

imatiach-msft commented:
gently ping @sethah @Lewuathe @thunterdb @WeichenXu123 @jkbradley @actuaryzhang @srowen could you please take a look?   I've merged with latest code. Thank you!

imatiach-msft committed:
updated version

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7713/
Test PASSed.

SparkQA commented:
**[Test build #102046 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102046/testReport)** for PR 17084 at commit [`793a284`](https://github.com/apache/spark/commit/793a28453286c8f6b26d7d52281147f05ace6514).

SparkQA commented:
**[Test build #102046 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102046/testReport)** for PR 17084 at commit [`793a284`](https://github.com/apache/spark/commit/793a28453286c8f6b26d7d52281147f05ace6514).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BinaryClassificationMetrics @Since(\"3.0.0\") (`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7802/
Test PASSed.

SparkQA commented:
**[Test build #102177 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102177/testReport)** for PR 17084 at commit [`e78076b`](https://github.com/apache/spark/commit/e78076bbb463b3aa5081a71c6e3a155b5e421e4a).

SparkQA commented:
**[Test build #102177 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102177/testReport)** for PR 17084 at commit [`e78076b`](https://github.com/apache/spark/commit/e78076bbb463b3aa5081a71c6e3a155b5e421e4a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BinaryClassificationMetrics @Since(\"3.0.0\") (`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102177/
Test PASSed.

holdenk commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryLabelCounter.scala:
```diff
@@ -24,14 +24,22 @@ package org.apache.spark.mllib.evaluation.binary
 private[evaluation] class BinaryLabelCounter(
-    var numPositives: Long = 0L,
-    var numNegatives: Long = 0L) extends Serializable {
+    var numPositives: Double = 0.0,
```
This is no longer the number of positive records necessarily, maybe update this variable name as well?

imatiach-msft replied:
done!  updated to weightedNumPositives and weightedNumNegatives

holdenk commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala:
```diff
@@ -27,33 +27,33 @@ private[evaluation] trait BinaryClassificationMetricComputer extends Serializabl
       0.0
     } else {
-      c.numTruePositives.toDouble / c.numPositives
+      c.weightedTruePositives.toDouble / c.weightedPositives
```
Is the call `toDouble` needed here still?

imatiach-msft replied:
done!  good point, removed

holdenk commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala:
```diff
@@ -27,33 +27,33 @@ private[evaluation] trait BinaryClassificationMetricComputer extends Serializabl
 private[evaluation] object Recall extends BinaryClassificationMetricComputer {
   override def apply(c: BinaryConfusionMatrix): Double = {
-    if (c.numPositives == 0) {
+    if (c.weightedPositives == 0) {
```
Should this be 0.0?

imatiach-msft replied:
done!

holdenk commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala:
```diff
@@ -27,33 +27,33 @@ private[evaluation] trait BinaryClassificationMetricComputer extends Serializabl
 private[evaluation] object FalsePositiveRate extends BinaryClassificationMetricComputer {
   override def apply(c: BinaryConfusionMatrix): Double = {
-    if (c.numNegatives == 0) {
+    if (c.weightedNegatives == 0) {
```
Should this be 0.0?

imatiach-msft replied:
done!

holdenk commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/binary/BinaryClassificationMetricComputers.scala:
```diff
@@ -27,33 +27,33 @@ private[evaluation] trait BinaryClassificationMetricComputer extends Serializabl
       0.0
     } else {
-      c.numFalsePositives.toDouble / c.numNegatives
+      c.weightedFalsePositives.toDouble / c.weightedNegatives
```
Is this toDouble still needed?

imatiach-msft replied:
done!  good point, not needed, removed

holdenk reviewed (commented):
I did a quick first pass review, thanks for working on this :)

imatiach-msft committed:
updated based on comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7842/
Test PASSed.

SparkQA commented:
**[Test build #102232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102232/testReport)** for PR 17084 at commit [`af2bca6`](https://github.com/apache/spark/commit/af2bca613f6f433ce5eb12a11dba84ef48771bf8).

SparkQA commented:
**[Test build #102232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102232/testReport)** for PR 17084 at commit [`af2bca6`](https://github.com/apache/spark/commit/af2bca613f6f433ce5eb12a11dba84ef48771bf8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102232/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7877/
Test PASSed.

SparkQA commented:
**[Test build #102274 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102274/testReport)** for PR 17084 at commit [`5de56a5`](https://github.com/apache/spark/commit/5de56a550170c71b41b646b582976270c4b4c36f).

imatiach-msft commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -41,12 +41,26 @@ import org.apache.spark.sql.DataFrame
-    @Since("1.3.0") val numBins: Int) extends Logging {
+class BinaryClassificationMetrics @Since("3.0.0") (
+    val numBins: Int,
+    @Since("3.0.0") val scoreAndLabelsWithWeights: RDD[(Double, (Double, Double))])
```
it looks like I need to revert this based on the other two PRs

imatiach-msft replied:
done!  made code more similar to other two merged PRs

imatiach-msft committed:
made code more similar to other two PRs

SparkQA commented:
**[Test build #102275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102275/testReport)** for PR 17084 at commit [`65f998c`](https://github.com/apache/spark/commit/65f998cfd8e99fdde366008219b4a04a90aa823e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7878/
Test PASSed.

SparkQA commented:
**[Test build #102275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102275/testReport)** for PR 17084 at commit [`65f998c`](https://github.com/apache/spark/commit/65f998cfd8e99fdde366008219b4a04a90aa823e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102275/
Test FAILed.

imatiach-msft committed:
fix MIMA

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7879/
Test PASSed.

SparkQA commented:
**[Test build #102276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102276/testReport)** for PR 17084 at commit [`9f6a5d5`](https://github.com/apache/spark/commit/9f6a5d58e62d636169da8fe8e6a1332c3b384548).

SparkQA commented:
**[Test build #102276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102276/testReport)** for PR 17084 at commit [`9f6a5d5`](https://github.com/apache/spark/commit/9f6a5d58e62d636169da8fe8e6a1332c3b384548).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102276/
Test FAILed.

SparkQA commented:
**[Test build #102274 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102274/testReport)** for PR 17084 at commit [`5de56a5`](https://github.com/apache/spark/commit/5de56a550170c71b41b646b582976270c4b4c36f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102274/
Test PASSed.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/BinaryClassificationEvaluator.scala:
```diff
@@ -68,21 +74,33 @@ class BinaryClassificationEvaluator @Since("1.4.0") (@Since("1.4.0") override va
-        case Row(rawPrediction: Double, label: Double) => (rawPrediction, label)
+    val scoreAndLabelsWithWeights =
+      dataset.select(col($(rawPredictionCol)), col($(labelCol)).cast(DoubleType),
+        if (!isDefined(weightCol) || $(weightCol).isEmpty) lit(1.0)
```
We might need to wrap this differently for clarity, like
```
dataset.select(
  col(...),
  ...
  if () else ...).rdd.map {
    case ...
```

imatiach-msft replied:
to confirm, you mean just the formatting needs to be changed? done!

srowen replied:
Yes just formatting

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -41,9 +41,18 @@ import org.apache.spark.sql.DataFrame
-    @Since("1.3.0") val scoreAndLabels: RDD[(Double, Double)],
-    @Since("1.3.0") val numBins: Int) extends Logging {
+class BinaryClassificationMetrics @Since("3.0.0") (
+    @Since("1.3.0") val scoreAndLabelsWithOptWeight: RDD[_ <: Product],
```
Let's see, so last time we decided this was OK because the type change is only visible at compile-time, and this should be source compatible. I don't _think_ the argument name change will be an issue here as it isn't a named optional arg. should be OK but we'll have to check with MiMa.

srowen replied:
BTW do we check elsewhere for positive weight? they need to be > 0?

imatiach-msft replied:
no we currently do not check for the weight > 0, should that live in BinaryClassificationMetrics?  I should probably update the other metrics classes as well then?

srowen replied:
There's no useful case for negative weights right? I can see allowing 0 for convenience, as it at least has a clear meaning.

Maybe this could be a separate change, but yeah anywhere the user supplies weights, it's useful to check this _somewhere_ along the way.

imatiach-msft replied:
Yes, I don't believe there is a useful case for negative weights but 0 may be useful.
I added a check in binary classification metrics.  Should I add to the other evaluators in this PR as well?

imatiach-msft replied:
specifically, the check was:
```
require(weight >= 0.0, s"instance weight, $weight has to be >= 0.0")
```

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -53,6 +62,15 @@ class BinaryClassificationMetrics @Since("1.3.0") (
+    * Retrieves the score and labels (for binary compatibility).
+    * @return The score and labels.
+    */
+  @Since("1.0.0")
```
Ah, was this visible as a getter? OK this is probably fine but just wasn't clear if it even became visible; is it used outside the constructor?

imatiach-msft replied:
If I understand correctly, a getter is constructed in binary only if the field is used somewhere else outside the class, and apparently it was used as a getter.  However, I can't tell where it was used from the current code, perhaps it was used that way in the past in an earlier version of the code.

imatiach-msft replied:
Actually, if that was the case in binary, why wouldn't it complain about the code right now - hmm, not quite sure why MIMA needs this.

srowen replied:
What if you remove it? right now I don't see it used anywhere outside the class. However we do need to retain the constructor; maybe that was the MiMa issue?

imatiach-msft replied:
yes, MIMA was complaining, and when I added this the error was fixed

srowen replied:
Hm OK, I guess it was a getter after all. It looks like it's since 1.3.0 though.

imatiach-msft replied:
done, updated Since to 1.3.0

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -53,6 +62,15 @@ class BinaryClassificationMetrics @Since("1.3.0") (
+    */
+  @Since("1.0.0")
+  def scoreAndLabels: RDD[(Double, Double)] = {
+    scoreLabelsWeight.map(values => (values._1, values._2._1))
```
`.map { case (prediction, (label, _)) => (prediction, label) }`?

imatiach-msft replied:
done!  great suggestion!

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -146,11 +164,13 @@ class BinaryClassificationMetrics @Since("1.3.0") (
+    // negatives within each bin, and then sort by score values in descending order.
+    val counts = scoreLabelsWeight.combineByKey(
+      createCombiner = (labelAndWeight: (Double, Double)) =>
+        new BinaryLabelCounter(0.0, 0.0) += (labelAndWeight._1, labelAndWeight._2),
```
Can you just init the counter with those values rather than add to 0?

imatiach-msft replied:
done!  great idea

imatiach-msft replied:
sorry, I just noticed this was causing the test failures and actually won't work - please see the class BinaryLabelCounter.scala.  The two values we init it with are actually weightedNumPositives and weightedNumNegatives.  When we do += (val1, val2), that operator is overloaded and it is actually taking a label and weight.  So they are not actually the same.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -146,11 +164,13 @@ class BinaryClassificationMetrics @Since("1.3.0") (
+      createCombiner = (labelAndWeight: (Double, Double)) =>
+        new BinaryLabelCounter(0.0, 0.0) += (labelAndWeight._1, labelAndWeight._2),
+      mergeValue = (c: BinaryLabelCounter, labelAndWeight: (Double, Double)) =>
+        c += (labelAndWeight._1, labelAndWeight._2),
```
Isn't this the same as `c += labelAndWeight`?

imatiach-msft replied:
I would think so but for some reason it didn't like that:
[error] /home/ilya/ilmat-spark/mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:173: overloaded method value += with alternatives:
[error]   (other: org.apache.spark.mllib.evaluation.binary.BinaryLabelCounter)org.apache.spark.mllib.evaluation.binary.BinaryLabelCounter <and>
[error]   (label: Double)org.apache.spark.mllib.evaluation.binary.BinaryLabelCounter
[error]  cannot be applied to ((Double, Double))
[error]         c += labelAndWeight,
[error]           ^

srowen replied:
Hm, confusing, because the current code is also passing a (Double, Double), but also, there doesn't seem to be a += method for (Double, Double) anyway. I'm not sure why this works! maybe some implicit conversion happening? weird.

imatiach-msft replied:
oh, there is a += operator overloaded for (Double, Double), see BinaryLabelCounter.scala (I added it in this PR - in addition to the += Label operator overload that already existed there).  Maybe having an operator overload that works on a tuple is confusing to add?  Perhaps I need to add an explicit method instead - but this seems cleaner and more in line with the existing += operator overload that takes the label as a Double.

imatiach-msft committed:
updated based on comments and fixed style

SparkQA commented:
**[Test build #102379 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102379/testReport)** for PR 17084 at commit [`ab7e51a`](https://github.com/apache/spark/commit/ab7e51a1bd0bfc90e9272f54e8d7edc2e43d96ec).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7966/
Test PASSed.

SparkQA commented:
**[Test build #102380 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102380/testReport)** for PR 17084 at commit [`d98980a`](https://github.com/apache/spark/commit/d98980a20645eacafaeb8868da216d1ed7f471de).

SparkQA commented:
**[Test build #102379 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102379/testReport)** for PR 17084 at commit [`ab7e51a`](https://github.com/apache/spark/commit/ab7e51a1bd0bfc90e9272f54e8d7edc2e43d96ec).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102379/
Test FAILed.

SparkQA commented:
**[Test build #102380 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102380/testReport)** for PR 17084 at commit [`d98980a`](https://github.com/apache/spark/commit/d98980a20645eacafaeb8868da216d1ed7f471de).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

imatiach-msft committed:
updated based on comments

imatiach-msft committed:
fixed failing tests

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #102510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102510/testReport)** for PR 17084 at commit [`955b022`](https://github.com/apache/spark/commit/955b0226e8614b5e024da2cc444a084429efae3f).

SparkQA commented:
**[Test build #102510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102510/testReport)** for PR 17084 at commit [`955b022`](https://github.com/apache/spark/commit/955b0226e8614b5e024da2cc444a084429efae3f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102510/
Test PASSed.

imatiach-msft commented:
@holdenk @srowen @actuaryzhang would you be able to take another look at this PR when you have a chance?  I've updated to latest, fixed all failing tests and I think I have fixed all of the comments.  Thank you!

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.sql.DataFrame
  * Evaluator for binary classification.
  *
- * @param scoreAndLabels an RDD of (score, label) pairs.
+ * @param scoreAndLabelsWithOptWeight an RDD of (score, label, weight) tuples.
```
It can also be a `(score, label)` tuple right?

imatiach-msft replied:
yes, good point, will update doc

imatiach-msft replied:
done!

srowen reviewed (commented):
It's looking OK; does Pyspark need an update, and does the multiclass evaluator need an update to match?

imatiach-msft commented:
@srowen multiclass weight columns PR is already merged:
https://github.com/apache/spark/pull/17086
The pyspark side will need to be updated.  Should that be part of this PR or a separate PR?

srowen commented:
Ah right, long since lost track. Yeah it would make sense to update it here too. I don't think R has an evaluator?

imatiach-msft committed:
updated doc based on comments

SparkQA commented:
**[Test build #102621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102621/testReport)** for PR 17084 at commit [`079e114`](https://github.com/apache/spark/commit/079e1140792cd4e3c5fb21ebbb1dd9aca370c977).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8172/
Test PASSed.

SparkQA commented:
**[Test build #102621 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102621/testReport)** for PR 17084 at commit [`079e114`](https://github.com/apache/spark/commit/079e1140792cd4e3c5fb21ebbb1dd9aca370c977).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102621/
Test PASSed.

imatiach-msft committed:
updated python code

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8252/
Test PASSed.

SparkQA commented:
**[Test build #102720 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102720/testReport)** for PR 17084 at commit [`a571adc`](https://github.com/apache/spark/commit/a571adcf5dcbcfb7fcd357fecbf74093df95d4aa).

SparkQA commented:
**[Test build #102720 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102720/testReport)** for PR 17084 at commit [`a571adc`](https://github.com/apache/spark/commit/a571adcf5dcbcfb7fcd357fecbf74093df95d4aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BinaryClassificationEvaluator(JavaEvaluator, HasLabelCol, HasRawPredictionCol, HasWeightCol,`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102720/
Test PASSed.

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetricsSuite.scala:
```diff
@@ -82,6 +82,34 @@ class BinaryClassificationMetricsSuite extends SparkFunSuite with MLlibTestSpark
+    val fpr = numFalsePositives.map(f => f / numNegatives)
+    val rocCurve = Seq((0.0, 0.0)) ++ fpr.zip(recalls) ++ Seq((1.0, 1.0))
+    val pr = recalls.zip(precisions)
+    val prCurve = Seq((0.0, 1.0)) ++ pr
```
You don't need to change this, just an aside, but you can write in cases like this `(0.0, 1.0) +: pr` I know the other code is written like this so don't change it.

imatiach-msft replied:
nice tip, thank you!

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetricsSuite.scala:
```diff
@@ -82,6 +82,34 @@ class BinaryClassificationMetricsSuite extends SparkFunSuite with MLlibTestSpark
+    val metrics = new BinaryClassificationMetrics(scoreAndLabelsWithWeights, 0)
+    val thresholds = Seq(0.8, 0.6, 0.4, 0.1)
+    val numTruePositives =
+      Seq(1.0 * w1, 1.0 * w1 + 2.0 * w2, 1.0 * w1 + 2.0 * w2, 3.0 * w2 + 1.0 * w1)
```
Also not a big deal but do you need to multiply by 1.0? if it's for clarity, OK. do they even need to be doubles?

imatiach-msft replied:
Yep, 1.0 is for clarity.  Changed to int instead of double.

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetricsSuite.scala:
```diff
@@ -82,6 +82,34 @@ class BinaryClassificationMetricsSuite extends SparkFunSuite with MLlibTestSpark
+    val precisions = numTruePositives.zip(numFalsePositives).map { case (t, f) =>
+      t.toDouble / (t + f)
+    }
+    val recalls = numTruePositives.map(t => t / numPositives)
```
Another tiny nit, you don't need to change this, but you can write `.map(_  / numPositives)`

imatiach-msft replied:
good idea, done!

srowen commented on mllib/src/test/scala/org/apache/spark/ml/evaluation/BinaryClassificationEvaluatorSuite.scala:
```diff
@@ -71,6 +71,33 @@ class BinaryClassificationEvaluatorSuite
+    val evaluator = new BinaryClassificationEvaluator()
+      .setMetricName("areaUnderROC").setWeightCol(weightCol)
+    val vectorDF = Seq(
+      (0d, Vectors.dense(2.5, 12), 1.0),
```
Nit: 0.0 instead of 0d. The "d" or "D" suffix I think might read for a moment like hex or something.

imatiach-msft replied:
done, also updated the test above which was also using 0d

srowen commented on python/pyspark/mllib/evaluation.py:
```diff
@@ -40,16 +40,28 @@ class BinaryClassificationMetrics(JavaModelWrapper):
 
-    def __init__(self, scoreAndLabels):
-        sc = scoreAndLabels.ctx
+    def __init__(self, scoreAndLabelsWithOptWeight):
```
Ah, this might be a problem in Python in a way it isn't in Scala. The name of the parameter matters. I think you can no longer call this with `...(scoreAndLabels=...)`. It's surprising if someone would do that, but, possible. I think at the end, maybe we don't change this parameter name to be safe, even though it now also accepts weights. That's not so bad.

And then I wonder, maybe change back the name of the parameter in Scala too? It's really scoreAndLabels(AndWeightsThoughThat'sMoreOfADetail)

imatiach-msft replied:
done, also updated the scala side and the multiclass metrics python and scala classes.  It looks like the python side for regression wasn't updated so I will do that in a separate PR and undo the scala parameter name change there too.

imatiach-msft committed:
updated based on new comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8265/
Test PASSed.

SparkQA commented:
**[Test build #102733 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102733/testReport)** for PR 17084 at commit [`d8a5865`](https://github.com/apache/spark/commit/d8a5865425805bb4064b65e621880456269894d5).

SparkQA commented:
**[Test build #102733 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102733/testReport)** for PR 17084 at commit [`d8a5865`](https://github.com/apache/spark/commit/d8a5865425805bb4064b65e621880456269894d5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MulticlassMetrics @Since(\"1.1.0\") (predictionAndLabels: RDD[_ <: Product]) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102733/
Test FAILed.

SparkQA commented:
**[Test build #102734 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102734/testReport)** for PR 17084 at commit [`00bfec1`](https://github.com/apache/spark/commit/00bfec119b2e15911fab616fe9f755d364cad3b2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8266/
Test PASSed.

SparkQA commented:
**[Test build #102734 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102734/testReport)** for PR 17084 at commit [`00bfec1`](https://github.com/apache/spark/commit/00bfec119b2e15911fab616fe9f755d364cad3b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MulticlassMetrics @Since(\"1.1.0\") (predictionAndLabels: RDD[_ <: Product]) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102734/
Test PASSed.

imatiach-msft commented:
@srowen I think I have fixed all pending comments and the tests are currently passing for this PR.  I couldn't find any evaluators in the R code, but then again I am not as familiar with sparkR (however, I've used sparklyr before https://github.com/rstudio/sparklyr).  Please let me know if there are any other comments that need to be addressed.  Thank you!

srowen commented:
Merged to master
